extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1393,67 +1393,69 @@ private Yielder<Object[]> getFinalResultsYielder(
+                                                    )
+                                                )
+                                            );
+                                          }
+                                          catch (IOException e) {
+                                            throw new RuntimeException(e);
+                                          }
+                                        }
+                                    ).collect(Collectors.toList())
+                   ).flatMap(
+                       frame -> {
+                         final Cursor cursor = FrameProcessors.makeCursor(
+                             frame,
+                             queryKernel.getStageDefinition(finalStageId).getFrameReader()
+                         );
+
+                         final ColumnSelectorFactory columnSelectorFactory = cursor.getColumnSelectorFactory();
+                         final ColumnMappings columnMappings = task.getQuerySpec().getColumnMappings();
+                         @SuppressWarnings("rawtypes")
+                         final List<ColumnValueSelector> selectors =
+                             columnMappings.getMappings()
+                                           .stream()
+                                           .map(
+                                               mapping ->
+                                                   columnSelectorFactory.makeColumnValueSelector(mapping.getQueryColumn())
+                                           ).collect(Collectors.toList());
+
+                         final List<SqlTypeName> sqlTypeNames = task.getSqlTypeNames();
+                         final List<Object[]> retVal = new ArrayList<>();
+                         while (!cursor.isDone()) {
+                           final Object[] row = new Object[columnMappings.size()];
+                           for (int i = 0; i < row.length; i++) {
+                             final Object value = selectors.get(i).getObject();
+                             if (sqlTypeNames == null || task.getSqlResultsContext() == null) {
+                               // SQL type unknown, or no SQL results context: pass-through as is.
+                               row[i] = value;
+                             } else {
+                               row[i] = SqlResults.coerce(
+                                   context.jsonMapper(),
+                                   task.getSqlResultsContext(),
+                                   value,
+                                   sqlTypeNames.get(i)
+                               );
+                             }
+                           }
+                           retVal.add(row);
+                           cursor.advance();
+                         }
+
+                         return Sequences.simple(retVal);
+                       }
+                   )
+                   .limit(Limits.MAX_SELECT_RESULT_ROWS)
```

Comment:
For the end user, we need to set a flag in the report mentioning the results are truncated.